### ISSUE TITLE
Run Background service Error

### ISSUE THREAD

**Issue by p3r4f089** [NativeScript/nativescript-geolocation#86]
### Which platform(s) does your issue occur on?
- Android version 
- emulator and device

### Please, provide the following version numbers that your issue occurs with:

- CLI: 3.2.1
- Cross-platform modules: 3.2.0
- Runtime(s): "tns-android": {
                           "version": "3.2.0"
                      }
- Plugin(s):
"dependencies": {
    "buffer": "^5.0.8",
    "js-base64": "^2.3.2",
    "moment": "^2.19.1",
    "native-socket": "file:./files/Nativesocket",
    "nativescript-android-utils": "^1.0.2",
    "nativescript-bootreceiver": "^1.0.0",
    "nativescript-checkbox": "^3.0.1",
    "nativescript-couchbase": "^1.0.18",
    "nativescript-geolocation": "4.2.0",
    "nativescript-loading-indicator": "^2.4.0",
    "nativescript-localize": "^2.0.1",
    "nativescript-mapbox": "^3.1.2",
    "nativescript-permissions": "^1.2.3",
    "nativescript-pulltorefresh": "^2.0.2",
    "nativescript-secure-storage": "^2.2.1",
    "nativescript-theme-core": "~1.0.2",
    "nativescript-toast": "^1.4.6",
    "netchecksum": "^0.1.1",
    "timrjs": "^1.0.1",
    "tns-core-modules": "^3.2.0",
    "tns-ios": "3.2.0",
    "validator": "^8.2.0"
  },
  "devDependencies": {
    "awesome-typescript-loader": "~3.1.3",
    "babel-traverse": "6.25.0",
    "babel-types": "6.25.0",
    "babylon": "6.17.4",
    "copy-webpack-plugin": "~4.0.1",
    "extract-text-webpack-plugin": "^3.0.1",
    "lazy": "1.0.11",
    "nativescript-css-loader": "~0.26.0",
    "nativescript-dev-sass": "^1.3.2",
    "nativescript-dev-typescript": "^0.5.1",
    "nativescript-dev-webpack": "^0.8.0",
    "nativescript-worker-loader": "~0.8.1",
    "raw-loader": "~0.5.1",
    "resolve-url-loader": "^2.1.1",
    "tns-platform-declarations": "^3.1.1",
    "typescript": "^2.5.3",
    "webpack": "~3.2.0",
    "webpack-bundle-analyzer": "^2.8.2",
    "webpack-sources": "~1.0.1"
  },
### Please, tell us how to recreate the issue in as much detail as possible. 

I have a service background and call watch or get position, with version 3.0.1 of the plugin well ok, but whit minors errors, I'am test with versions 4.1.0 and 4.2.0 and 
It works very well, so when app is open everything work ok, but when app is close and de service run on background mode i have a error, I'm using 

android.app.Service.extend("com.tns.nativescript", {
	onStartCommand: function(intent, flags, startId) {
		this.super.onStartCommand(intent, flags, startId);
		return android.app.Service.START_STICKY;
	},
	onCreate: function() {			
        watchLocation(); // here call wathc or get position			
	},
	onBind: function(intent) {
		console.log("on Bind Services");
	},
	onUnbind: function(intent){
		console.log('UnBind Service');
	},
  	onDestroy: function(){	
		console.log('service onDestroy);
    }
});

Error:

Error: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.content.Context.getPackageManager()' on a null object reference
                                                      com.google.android.gms.common.zzo.isGooglePlayServicesAvailable(Unknown Source)
                                                      com.google.android.gms.common.zze.isGooglePlayServicesAvailable(Unknown Source)
                                                      com.google.android.gms.common.GoogleApiAvailability.isGooglePlayServicesAvailable(Unknown Source)
                                                      com.tns.Runtime.callJSMethodNative(Native Method)
                                                      com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1084)
                                                      com.tns.Runtime.callJSMethodImpl(Runtime.java:966)
                                                      com.tns.Runtime.callJSMethod(Runtime.java:953)
                                                      com.tns.Runtime.callJSMethod(Runtime.java:937)
                                                      com.tns.Runtime.callJSMethod(Runtime.java:929)
                                                      com.tns.gen.java.lang.Runnable.run(Runnable.java:10)
                                                      android.os.Handler.handleCallback(Handler.java:739)
                                                      android.os.Handler.dispatchMessage(Handler.java:95)
                                                      android.os.Looper.loop(Looper.java:148)
                                                      android.app.ActivityThread.main(ActivityThread.java:5417)
                                                      java.lang.reflect.Method.invoke(Native Method)
                                                      com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                      com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

**Comment by p3r4f089** [comment]
when the app is running in the background works as expected, the problem starts when i kill the app , the service keeps running but the call to the location method gets an error as shown above.

**Comment by p3r4f089** [comment]
Hi @DimitarTachev I used the background service demo and it worked correctly, but when i try to get the position outside of the service it fails.


my code:

```
import * as geolocation from "nativescript-geolocation";
import { Accuracy } from "ui/enums";
import * as utils from "utils/utils";
declare let com: any;

var context = utils.ad.getApplicationContext();
var intent = new android.content.Intent(context, com.mc.TrackService.class);

let Toast = require("nativescript-toast");
let watchId;

(<any>android.app.Service).extend("com.mc.TrackService", {
    onStartCommand: function (intent, flags, startId) {
        this.super.onStartCommand(intent, flags, startId);
        return android.app.Service.START_STICKY;
    },
    onCreate: function () {
        geolocation.enableLocationRequest().then(function () {
            watchPosition();
        }, function (e) {
            console.log("Background enableLocationRequest error: " + (e.message || e));
        });
    },
    onBind: function (intent) {
        console.log("on Bind Services");
    },
    onUnbind: function(intent){
        console.log('UnBind Service');
    },
    onDestroy: function() {
        console.log('service onDestroy');
        clearWatch();
    }
});

function getCurrent(){
    geolocation.getCurrentLocation({
        desiredAccuracy: Accuracy.high,
        maximumAge: 5000,
        timeout: 10000
    })
        .then(function (loc) {
            if (loc) {
                console.log('My CurrentPosition :D ' + loc);
            }
        }, function (e) {
            console.error("Error: " + (e.message || e));
        });
}


function watchPosition(){
    watchId = geolocation.watchLocation(
        function (loc) {
            if (loc) {
                let toast = Toast.makeText('Background Location: ' + loc.latitude + ' ' + loc.longitude);
                toast.show();
                console.log('Background Location: ' + loc.latitude + ' ' + loc.longitude);
            }
        },
        function (e) {
            console.error("Background watchLocation error: " + (e.message || e));
        },
        {
            desiredAccuracy: Accuracy.high,
            updateDistance: 0.1,
            updateTime: 3000,
            minimumUpdateTime: 100
        });
}

function clearWatch(){
    if(watchId != null){
        geolocation.clearWatch(watchId);
    }
    
}
```



and Error

![captura de pantalla 2017-11-02 a la s 11 55 12 a m](https://user-images.githubusercontent.com/7819874/32339668-7daae384-bfc6-11e7-8570-5924ec8725ea.png)

**Comment by agrw** [comment]
getting same error, some google play service available check messed up?? 

![screenshot_20171102-142311](https://user-images.githubusercontent.com/5850889/32343821-3d36a292-bfdb-11e7-8423-4d40cdec1d17.png)

**Comment by agrw** [comment]
Error: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.content.Context.getPackageManager()' on a null object reference
    com.google.android.gms.common.zzo.isGooglePlayServicesAvailable(Unknown Source)
    com.google.android.gms.common.zze.isGooglePlayServicesAvailable(Unknown Source)
    com.google.android.gms.common.GoogleApiAvailability.isGooglePlayServicesAvailable(Unknown Source)
    com.tns.Runtime.callJSMethodNative(Native Method)
    com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1084)
    com.tns.Runtime.callJSMethodImpl(Runtime.java:966)
    com.tns.Runtime.callJSMethod(Runtime.java:953)
    com.tns.Runtime.callJSMethod(Runtime.java:937)
    com.tns.Runtime.callJSMethod(Runtime.java:929)
    com.tns.gen.java.lang.Runnable.run(Runnable.java:10)
    android.os.Handler.handleCallback(Handler.java:739)
    android.os.Handler.dispatchMessage(Handler.java:95)
    android.os.Looper.loop(Looper.java:148)
    android.app.ActivityThread.main(ActivityThread.java:7325)
    java.lang.reflect.Method.invoke(Native Method)
    com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
    com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)

**Comment by agrw** [comment]
For nativescript-gelocation plugin, if I change line 194 in file geolocation.android.js  from 
var resultCode = GoogleApiAvailability.isGooglePlayServicesAvailable( application_1.android.foregroundActivity);   
to
var resultCode = googleApiAvailability.isGooglePlayServicesAvailable(application_1.android.context);

Crash does not occur, the  "application_1.android.foregroundActivitity"  is not defined for background services? Whereas application_1.android.context is always defined. 
Is this fix correct?

**Comment by zbranzov** [comment]
@p3r4f089 Using the demo app I'm unable to reproduce the error. I'm getting position using the Get Current Location button which is outside the service. Can you elaborate a bit more and provide a complete example? Maybe I miss something.

**Comment by agrw** [comment]
I'm getting the current location inside the service, the variable application_1.android.foregroundActivity is null inside the service, but it you use application_1.android.context instead in call to  googleApiAvailability.isGooglePlayServicesAvailable everything works. on line 194 in geolocation.android.js  file. With this change everything works fine inside service or in foreground of app.

**Comment by Dragons0458** [comment]
@agrw Thank you! This works! Administrators must upload this change to the repository

**Comment by lini** [comment]
`foregroundActivity -> context` fix is published in v4.2.1

**Comment by marcossaore** [comment]
I want to do the same, leave application in the background and send the location to a webservice, does it work? Can you help me?
